File: src/foam/nanos/logger/GUIDLogger.java
/**
 * @license
 * Copyright 2018 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */
package foam.nanos.logger;

import java.util.Date;
import java.util.Random;

import foam.core.*;
import foam.dao.DAO;

public class GUIDLogger
  extends AbstractLogger
{
  private DAO logDAO_ = null;
  private Random r = new Random();
  public GUIDLogger(X x) {
    logDAO_ = (DAO) x.get("logDAO");
  }

  private FObject generateLogModel(String type, Object... args) {
    Date d = new Date();
    Log log = new Log();
    log.setId(Long.toString(d.getTime()) + "-" + Integer.toString(r.nextInt(100000), 5));
    log.setTime(new Date());
    log.setFrom("BackEnd");
    log.setBelong(System.getProperty("user.name"));
    log.setType(type);
    log.setDescription(args[0].toString());
    log.setDetail(combine(args));
    return log;
  }

  public void log(Object... args) {
    logDAO_.put(generateLogModel("log", args));
  }

  public void info(Object... args) {
    logDAO_.put(generateLogModel("info", args));
  }

  public void warning(Object... args) {
    logDAO_.put(generateLogModel("warning", args));
  }

  public void error(Object... args) {
    logDAO_.put(generateLogModel("error", args));
  }

  public void debug(Object... args) {
    logDAO_.put(generateLogModel("debug", args));
  }
}

File: src/foam/nanos/logger/GUIDLogger.js
/**
 * @license
 * Copyright 2018 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */
foam.CLASS({
  package: 'foam.nanos.logger',
  name: 'GUIDLogger',

  imports: [
    'logDAO',
    'user'
  ],

  requires: [
    'foam.nanos.logger.Log'
  ],

  properties: [
    {
      name: 'self',
      factory: function(){
        return this;
      }
    }
  ],

  methods: [
    function outputLogger(type, array) {
      if (array.length > 1) {
        var ret = ''
        for (var i = 0 ; i < array.length ; i++ ) {
          ret = ret + this.normalizeDetail(array[i]);
        }
        this.addToDAO(type, 'Please see details', ret);
      } else if ( array.length == 1 ) {
        var des = this.normalizeDescription(array[0]) || 'Please see details';
        this.addToDAO(type, des, this.normalizeDetail(array[0]));
      }
    },
    function log() {
      this.outputLogger('log', Array.from(arguments))
    },
    function warning() {
      this.outputLogger('warning', Array.from(arguments))
    },
    function info() {
      this.outputLogger('info', Array.from(arguments))
    },
    function error() {
      this.outputLogger('error', Array.from(arguments))
    },
    function debug() {
      this.outputLogger('debug', Array.from(arguments))
    },
    function normalizeDetail(e) {
      if ( foam.core.Exception.isInstance(e) ) {
        return '' + e + '\n' + '[name]: ' + e.name + '\n' + '[message]: ' + e.message + '\n';
      } else if ( e instanceof Error ) {
        return '' + e + '\n' + '[name]: ' + e.name + '\n' + '[message]: ' + e.message + '\n';
      } else {
        return '' + e + '\n';
      }
    },
    function normalizeDescription(e) {
      return (!! e.name) ? '' + e.name : undefined; 
    },
    function addToDAO(type, description, detail) {
      var l = this.generateLogModel(type, description, detail);
      this.logDAO.put(l);
    },
    function generateLogModel(type, description, detail) {
      return this.Log.create({
        time: new Date(),
        from: 'Web',
        user: (! this.user) ? '' : '' + this.user.firstName + ' ' + this.user.lastName,
        type: type,
        description: description,
        detail: detail
      })
    }
  ]
})